Add Jinja support

ShaheedHaque opened the pull request:
## Description

To reproduce the results to date requires the following example changes on the Django side...

1. Configure template files ending in ".jinja" to be processed via Jinja:
    ```python
        TEMPLATES = [
        {
            "BACKEND": "django_jinja.backend.Jinja2",  # Jinja backend
            'DIRS': [os.path.join(BASE_DIR, 'templates')],
            "APP_DIRS": True,
            "OPTIONS": {
                'context_processors': [
                    'django.contrib.auth.context_processors.auth',
                    'django.contrib.messages.context_processors.messages',
                    ...
                ],
                "match_extension": (".jinja"),  # suffix for Jinja templates
                "app_dirname": "templates",
    
                # Can be set to "jinja2.Undefined" or any other subclass.
                "undefined": None,
    
                "newstyle_gettext": True,
                "tests": {
                    # "mytest": "path.to.my.test",
                },
                "filters": {
                    # "myfilter": "path.to.my.filter",
                },
                "globals": {
                    # "myglobal": "path.to.my.globalfunc",
                },
                "constants": {
                    # "foo": "bar",
                },
                "extensions": [
                    "jinja2.ext.do",
                    "jinja2.ext.loopcontrols",
                    "jinja2.ext.i18n",
                    "django_jinja.builtins.extensions.CsrfExtension",
                    "django_jinja.builtins.extensions.CacheExtension",
                    "django_jinja.builtins.extensions.TimezoneExtension",
                    "django_jinja.builtins.extensions.UrlsExtension",
                    "django_jinja.builtins.extensions.StaticFilesExtension",
                    "django_jinja.builtins.extensions.DjangoFiltersExtension",
                    "reactpy_django.templatetags.jinja.ReactPyExtension",    # new extension
                ],
                "bytecode_cache": {
                    "name": "default",
                    "backend": "django_jinja.cache.BytecodeCache",
                    "enabled": False,
                },
                "autoescape": True,
                "auto_reload": DEBUG,
                "translation_engine": "django.utils.translation",
            },
        },
        {
            'BACKEND': 'django.template.backends.django.DjangoTemplates', # Django templates.

    ```
2. Add the new app:
    ```python
    INSTALLED_APPS = [
         ...
        "reactpy_django",
     ]

    ```
3. File of components in `components.py`:
    ```python
    from reactpy import component, html

    @component
    def hello_world(recipient: str):
        return html.h1(f"Hello {recipient}!")
    ```
4. Test view in `views/react.py`:
    ```python
    from django.shortcuts import render

    def jinja_view(request):
        return render(request, "my-template.html.jinja")
    ```
6. Template for the view `templates/my-template.html.jinja`:
    ```html
    <!DOCTYPE html>
    <html>
    <body>
    {{ component("paiyroll.components.hello_world", recipient="World") }}
    </body>
    </html>
    ```
8. `project/asgi.py`
    ```python
    from reactpy_django import REACTPY_WEBSOCKET_ROUTE
 
    application = ProtocolTypeRouter({
      'http': get_asgi_application(),
      'websocket': AuthMiddlewareStack(
          URLRouter(
            ... + [REACTPY_WEBSOCKET_ROUTE]
         )
    ``` 

9. And finally `project/urls.py`:
    ```python
    from ...views import react
 
    urlpatterns = [
         ...
         path("reactpy/", include("reactpy_django.http.urls")),
         path("react/", react.jinja_view),
     ]
     ```
Once Django is restarted, navigating to the view, you should see
    ```
    Hello World!
    ```




## Checklist:

Please update this checklist as you complete each item:

-   [ ] Tests have been developed for bug fixes or new functionality.
-   [ ] The changelog has been updated, if necessary.
-   [ ] Documentation has been updated, if necessary.
-   [ ] GitHub Issues closed by this PR have been linked.

<sub>By submitting this pull request you agree that all contributions comply with this project's open source license(s).</sub>

ShaheedHaque committed:
WIP Jinja support.

ShaheedHaque commented:
OK, this now render the "Hello World!" correctly.

Feedback requested before thinking about tests/docs etc.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+from jinja2.runtime import Context, Undefined
+
+from .reactpy import component as djt_component
+from .. import config
```
No relative imports

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+from jinja2.ext import Extension
+from jinja2.runtime import Context, Undefined
+
+from .reactpy import component as djt_component
```
Use `import reactpy` instead, then call the component as `reactpy.component`

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+        Returns:
+            Whatever the components returns.
+        """
+        djt_context = RequestContext(__context.parent['request'], autoescape=__context.eval_ctx.autoescape)
```
Don't use acronyms when possible, to help future contributors. This variable should be named `request_context`

Archmonger reviewed (commented):
A few comments since I took a early peak at the draft.

Archmonger commented:
I don't think we should be storing components within `environment.globals["component"]`. That is both unnecessary memory overhead, and won't be multi-threading/multi-processing safe.

I would suggest the following interface instead:

`{{ component("dotted.path.to.component", args=[variable]) }}`

Also I know that Django template tags can be treated like normal functions, so we might want to call it directly within this Jinja tag if possible. That would save us the burden of keeping parity between their interfaces.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+"""
+Jinja support.
+"""
+import typing as t
```
Import whatever you need `from typing import ...` instead of abstract import renaming.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+
+        {{ component(*args, **kwargs) }}
+    """
+    DJT_TEMPLATE = 'reactpy/component.html'
```
No point as storing this as a class attribute if it's only used once

ShaheedHaque replied:
This is really a reminder to pull this out into somewhere as thew value is required to be the same as used in the decorator for templatetags.component.

ShaheedHaque replied:
@Archmonger I think this is the last remaining unresolved item. Any thoughts on sharing this value, or shall I just go ahead and hard code it?

Archmonger replied:
Let's make it a global within the Django template tag module for now.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+        environment.globals["component"] = self._jinja_component
+
+    @pass_context
+    def _jinja_component(self, __context: Context, dotted_path: str, *args: t.Any, host: str | None = None,
```
May as well just accept `_jinja_component(self, __context: Context, *args, **kwargs)`.

That is, unless there's some tools out there for Jinja auto-complete within templates.

On that note, this method should be renamed to `def template_tag( ... )`.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+            Whatever the components returns.
+        """
+        djt_context = RequestContext(__context.parent['request'], autoescape=__context.eval_ctx.autoescape)
+        context = djt_component(djt_context, dotted_path, *args, host=host, prerender=prerender, **kwargs)
```
Rename to `django_template_tag = ...`

ShaheedHaque replied:
Not sure what you mean, see updated code?

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,67 @@
+        #
+        # TODO: can this be usefully cached?
+        #
+        result = loader.render_to_string(self.DJT_TEMPLATE, context, __context.parent['request'])
+        return result
```
Don't assign variables to something that is directly returned.

ShaheedHaque committed:
Address second pass review comments.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,63 @@
+        environment.globals["component"] = self.template_tag
+
+    @pass_context
+    def template_tag(self, jinja_context: Context, dotted_path: str, *args, host: str | None = None,
+                     prerender: str = str(config.REACTPY_PRERENDER), **kwargs) -> str:
```
In order to maintain interface parity more easily, the params for this can be changed to

`(self, jinja_context: Context, *args, **kwargs)`

ShaheedHaque replied:
Yes, but then the structural similarity with the native code would be lost, and some annoying/ugly unpacking would be required on line 59. Surely that is worth avoiding?

Archmonger replied:
The alternative is I would need to start manually maintaining interface parity between Jinja and Django every time I add a new template tag feature, which is a royal pain.

ShaheedHaque replied:
Fair enough. One last thought...since `dotted_path` is required, how would you feel about:
```
def template_tag(self, jinja_context: Context, dotted_path: str, *args, **kwargs)
```

Archmonger replied:
Yeah that works.

ShaheedHaque committed:
Address lastoutstadnngcomment, and run black.

ShaheedHaque commented:
Resolved last outstanding comment. And ran black.

If we are to proceed, I'd propose to squash the current commits before thinking about tests/packaging/docs etc. (FWIW, in that regard, I assume the appropriate packaging would be as something like "reactpy_django[jinja]"...but I'm not currently sure how to go about implementing that).

Archmonger commented:
Optional dependecies can be defined within the package definition in [setup.py](https://github.com/reactive-python/reactpy-django/blob/6f79c4cf7ed8b0b4cd048817aa90ab12997caabc/setup.py#L47-L81). For example:

```python
{
  "extras_require": {
    "encryption": ["cryptography", "pycryptodome"],
  },
  ...
}
```

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,66 @@
+#
+# Point to our non-Django analogue.
+#
+DJT_TEMPLATE = "reactpy/component.html"
```
This does not solve the code duplication issue.

This should be stored within `reactpy_django.templatetags.reactpy` and imported into this file.

References to `"reactpy/component.html"` within `reactpy_django.templatetags.reactpy` will need to be replaced with this global.

ShaheedHaque replied:
Ah, I misunderstood what you wanted. Getting rid of the duplication was my point too.

Archmonger commented on src/reactpy_django/templatetags/jinja.py:
```diff
@@ -0,0 +1,66 @@
+# Copyright © 2023 Innovatie Ltd. All rights reserved.
```
All code within this repository must be encompassed by our repository's [copyright licensing](https://github.com/reactive-python/reactpy-django/blob/main/LICENSE.md), as described in the pull request template.

Having your company retain copyright for this one file will not be acceptable. This would mean that legally we would have no right to use, copy, distribute, or modify your contributions.

ShaheedHaque commented:
> Optional dependencies can be defined within the package in [setup.py](https://github.com/reactive-python/reactpy-django/blob/6f79c4cf7ed8b0b4cd048817aa90ab12997caabc/setup.py#L47-L81). For example:
> 
> ```python
> {
>   "extras_require": {
>     "encryption": ["cryptography", "pycryptodome"],
>   },
>   ...
> }
> ```

Thanks, I'm more or less familiar with this bit. I was more thinking about the new file I have added (and any supporting test file), and whether that could/should be omitted unless "[jinja] was specified?

ShaheedHaque committed:
Address copyright, and use of constant to locate template.

Archmonger commented:
I have some changes in one of my PRs that will simplify test configuration
https://github.com/reactive-python/reactpy-django/pull/190

TLDR: Our test suite can now run multiple different Django settings.py files.

Archmonger commented:
My PR has been merged. All settings files [within the test app](https://github.com/reactive-python/reactpy-django/tree/main/tests/test_app) will now get the test suite automatically run against them.

There shouldn't be anything blocking this PR anymore, so you'll need to do the following:

1. Add your dependencies to [test-env.txt](https://github.com/reactive-python/reactpy-django/blob/main/requirements/test-env.txt)
2. Add the needed dependencies as [optional extras](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies) using the `[jinja]` keyword.
3. Cop [settings_single_db.py](https://github.com/reactive-python/reactpy-django/blob/main/tests/test_app/settings_single_db.py) and modifying it to fit your test case.
4. Write some [conditional tests](https://stackoverflow.com/questions/38442897/how-do-i-disable-a-test-using-pytest) within the [test folder](https://github.com/reactive-python/reactpy-django/tree/main/tests/test_app/tests)
5. Write a new page of documentation based on [the current installation page](https://github.com/reactive-python/reactpy-django/blob/main/docs/src/learn/add-reactpy-to-a-django-project.md).

ShaheedHaque commented:
Noted. I've a full plate right now but have this on my radar.

Commit:
Add Jinja2 template support for ReactPy components

- Created `ReactPyExtension` Jinja2 extension with `component`,
  `pyscript_component`, and `pyscript_setup` globals that delegate to the
  existing Django template tag implementations.
- Refactored `reactpy.py` to export template name constants
  (COMPONENT_TEMPLATE, PYSCRIPT_COMPONENT_TEMPLATE, PYSCRIPT_SETUP_TEMPLATE)
  so they can be reused by the Jinja2 extension.
- Added JINJA_COMPONENT_REGEX to utils.py's RootComponentFinder so that
  {{ component(...) }} syntax in Jinja2 templates is auto-detected.
- Added test infrastructure: settings_jinja.py, jinja_env.py, jinja_views.py,
  jinja_urls.py, Jinja2 template files, and test_jinja.py.
- Made test_app/__init__.py gracefully handle missing `bun` binary by
  skipping the JS rebuild when artifacts already exist.
- Added jinja2 to pyproject.toml hatch-test extra-dependencies.
- Updated CHANGELOG.md.

(cherry picked from commit 3923cee0aa8f37a0346e59e05bf042275d32ca0f)

Commit:
Merge remote-tracking branch 'origin/main' into srh_jinja_support

Archmonger committed:
fix type and lint warnings

Commit:
Fix Jinja2 support review issues

- Remove invalid 'jinja2' entry from INSTALLED_APPS in settings_jinja.py
  (jinja2 is a Python library, not a Django app)
- Consolidate imports from reactpy_django.templatetags.reactpy into
  a single import block in jinja.py
- Add [project.optional-dependencies] jinja section to pyproject.toml
  so 'pip install reactpy_django[jinja]' works for end users
- Strengthen test_jinja.py assertions and remove empty view test stubs
  that provided no coverage

Commit:
Fix formatting and lint issues for Jinja2 support

- Run ruff format fix on test_jinja.py (single-line .from_string call)
- Run ruff check --fix on jinja.py (sorted imports per project convention)

Commit:
Fix Jinja2 _component prerender default value

The Jinja wrapper used prerender="" as default, but str_to_bool("")
raises ValueError. Fix by matching the Django tag behavior — default to
str(REACTPY_PRERENDER) when no explicit prerender value is provided.

Commit:
Fix test assertion in test_jinja_component_renders_without_error

The component template does not use data-reactpy attribute.
Check for mountComponent and the component dotted path instead.

Commit:
Remove optional jinja2 dependency, add import guard, add Jinja docs page

- Remove [project.optional-dependencies] jinja = ["jinja2"] from pyproject.toml
- Add try/except ModuleNotFoundError guard at the top of jinja.py with
  a helpful error message telling users to install jinja2
- Create docs/src/learn/add-reactpy-to-a-django-project-with-jinja.md
  with Jinja2-specific setup instructions
- Update mkdocs.yml nav to include the new Jinja page under Get Started

Commit:
Add env, jinja, jinja2, myproject, and vs to spellcheck dictionary

Archmonger committed:
docs cleanup

Archmonger committed:
add more robust checks
